### PR TITLE
Only spawn runner if behaviors for the language are present

### PR DESCRIPTION
**Pull request by leudz**
## 🌟 What is the purpose of this PR?

We used to spawn both Javascript and Python runners. https://github.com/hashintel/hash/pull/546 makes the Python runner only spawn when it's required.
This PR expand this to Javascript and Rust.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1199548034582004/1202252324296540/f) _(internal)_

## 🔍 What does this change?

- Create the Rust runner but only if not spawned
- Spawn the JS and Rust runner only when needed

## 📜 Does this require a change to the docs?

No

## 🛡 What tests cover this?

All tests using JS behaviors will check that the JS runner is correctly spawned.
I don't think there is any test to check it is not spawned.
I don't think there is any test for the Rust runner as it's not implemented yet.

## ❓ How to test this?

1.  Checkout the branch / view the deployment
2.  Try `cargo test`
3.  Confirm that all tests pass

### CODE REVIEW

File: packages/engine/src/proto.rs [hunk 1]
```diff
@@ -226,9 +227,27 @@ pub struct ExperimentRunBase {
...
+            rust: self.requires_runner(Language::Rust),
+            javascript: self.requires_runner(Language::JavaScript),
+        }
+    }
+
+    /// Returns `true` if the experiment uses the language's init or has any behavior of the
+    /// language.
+    fn requires_runner(&self, language: Language) -> bool {
+        let requires_init = match (language, &self.project_base.package_init.initial_state.name) {
+            (Language::JavaScript, InitialStateName::InitJson) => true,
```
Comment (leudz) [inline packages/engine/src/proto.rs:244]: I'm not sure if `InitJson` makes the JS runner required, I guessed it was.

File: packages/engine/src/proto.rs [hunk 2]
```diff
@@ -226,9 +227,27 @@ pub struct ExperimentRunBase {
...
+            python: self.requires_runner(Language::Python),
+            rust: self.requires_runner(Language::Rust),
+            javascript: self.requires_runner(Language::JavaScript),
+        }
+    }
+
+    /// Returns `true` if the experiment uses the language's init or has any behavior of the
+    /// language.
+    fn requires_runner(&self, language: Language) -> bool {
+        let requires_init = match (language, &self.project_base.package_init.initial_state.name) {
```
Comment (TimDiekmann) [inline packages/engine/src/proto.rs:243]: `match` is far more readable IMO, let's ignore clippy here.

File: packages/engine/src/proto.rs [hunk 3]
```diff
@@ -226,9 +227,27 @@ pub struct ExperimentRunBase {
 }
 
 impl ExperimentRunBase {
-    /// Returns `true` if the experiment uses Python init or has any Python behavior.
-    pub fn requires_python_runner(&self) -> bool {
-        self.project_base.package_init.initial_state.name == InitialStateName::InitPy
+    /// Returns a `RunnerSpawnConfig` matching the config required by the files present in the
+    /// experiment.
+    pub fn runner_spawn_config(&self) -> RunnerSpawnConfig {
```
Comment (TimDiekmann) [inline packages/engine/src/proto.rs:232]: No strong opinion, but I would have chosen `create_runner_spawn_config`. Feel free to ignore

File: packages/engine/src/proto.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/engine/src/proto.rs:245]: My guess is that's not required. Could you make a commit removing this line and check if CI fails?

File: packages/engine/src/proto.rs [hunk 4]
```diff
@@ -226,9 +227,27 @@ pub struct ExperimentRunBase {
 }
 
 impl ExperimentRunBase {
-    /// Returns `true` if the experiment uses Python init or has any Python behavior.
-    pub fn requires_python_runner(&self) -> bool {
-        self.project_base.package_init.initial_state.name == InitialStateName::InitPy
+    /// Returns a `RunnerSpawnConfig` matching the config required by the files present in the
```
Comment (TimDiekmann) [inline packages/engine/src/proto.rs:230]:
```suggestion
    /// Returns a [`RunnerSpawnConfig`] matching the config required by the files present in the
```

File: packages/engine/src/proto.rs (hunk 2 above)
Comment (leudz) [reply packages/engine/src/proto.rs:244]: :smile: I was checking how it looked when I saw your comment.
Ingored clippy https://github.com/hashintel/hash/pull/552/commits/531df30ad1795a5a79ec23c75cca3a2b0fcba25e

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
Could you adjust `sync_runners` as well?

File: packages/engine/src/proto.rs (hunk 3 above)
Comment (leudz) [reply packages/engine/src/proto.rs]: Renamed in https://github.com/hashintel/hash/pull/552/commits/f28e23cc9b846b9740a3f4c009d8bdbba20dd961

File: packages/engine/src/proto.rs (hunk 1 above)
Comment (leudz) [reply packages/engine/src/proto.rs]: CI is ok with it https://github.com/hashintel/hash/pull/552/checks?sha=b738148d03fe60e44e72bf84b1df9de264e54ace

**Review by TimDiekmann — APPROVED** [review]
🚀

**Pull request merged**